packages/toolkit/src/combineSlices.ts

```
@@ -0,0 +1,501 @@
+     * @returns original, unproxied state
+     * @throws if value passed is not a state Proxy
+     */
+    original: <
+      State extends CombinedSliceState<
+        StaticState,
+        LazyLoadedState,
+        InjectedKeys
+      >
+    >(
+      state: State
+    ) => WithOptionalProp<State, InjectedKeys>
+  }
+}
+
+type StaticState<Slices extends Array<AnySlice | AnyApi | ReducerMap>> =
+  UnionToIntersection<
+    Slices[number] extends infer Slice
+      ? Slice extends AnySlice
+        ? WithSlice<Slice>
+        : Slice extends AnyApi
+        ? WithApi<Slice>
+        : StateFromReducersMapObject<Slice>
+      : never
+  >
+
+const isSlice = (
+  maybeSlice: AnySlice | AnyApi | ReducerMap
+): maybeSlice is AnySlice =>
+  'actions' in maybeSlice && typeof maybeSlice.actions === 'object'
+
+const isApi = (maybeApi: AnySlice | AnyApi | ReducerMap): maybeApi is AnyApi =>
+  'reducerPath' in maybeApi && typeof maybeApi.reducerPath === 'string'
+
+const getReducers = (slices: Array<AnySlice | AnyApi | ReducerMap>) =>
+  slices.flatMap((sliceOrMap) =>
+    isSlice(sliceOrMap)
+      ? [[sliceOrMap.name, sliceOrMap.reducer] as const]
+      : isApi(sliceOrMap)
+      ? [[sliceOrMap.reducerPath, sliceOrMap.reducer] as const]
+      : Object.entries(sliceOrMap)
+  )
+
+const ORIGINAL_STATE = Symbol()
+
+const isStateProxy = (value: any) => !!value && !!value[ORIGINAL_STATE]
+
+const createStateProxy = <State extends object>(
+  state: State,
+  reducerMap: Partial<Record<string, Reducer>>
+) =>
+  new Proxy(state, {
+    get: (target, prop, receiver) => {
+      if (prop === ORIGINAL_STATE) return target
+      const result = Reflect.get(target, prop, receiver)
+      if (typeof result === 'undefined') {
+        const reducer = reducerMap[prop.toString()]
+        if (reducer) {
+          // ensure action type is random, to prevent reducer treating it differently
+          return reducer(undefined, { type: nanoid() })
```

do we want to throw an error if _this_ returns undefined?